src/tools/info/proto_info.c

```diff
@@ -134,32 +134,155 @@ ucs_status_t wait_completion(ucp_worker_h worker, ucs_status_ptr_t status_ptr)
     return status;
 }
 
+ucs_status_t
+create_listener(ucp_worker_h worker, ucp_listener_h *listener_p,
+                uint16_t *listen_port_p, ucp_ep_h *server_ep_p)
+{
+    ucp_listener_h listener;
+    struct sockaddr_in listen_saddr;
+    ucp_listener_params_t listen_params;
+    ucp_listener_attr_t listen_attr;
+    ucs_status_t status;
+    
+    set_listen_addr(&listen_saddr);
+
+    listen_params.field_mask         = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                                       UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER;
+    listen_params.sockaddr.addr      = (const struct sockaddr*)&listen_saddr;
+    listen_params.sockaddr.addrlen   = sizeof(listen_saddr);
+    listen_params.accept_handler.cb  = listener_accept_callback;
+    listen_params.accept_handler.arg = server_ep_p;
+
+    status = ucp_listener_create(worker, &listen_params, &listener);
+    if (status != UCS_OK) {
+        printf("<Failed to create UCP listener>\n");
+        goto out;
+    }
+
+    listen_attr.field_mask = UCP_LISTENER_ATTR_FIELD_SOCKADDR;
+
+    status = ucp_listener_query(listener, &listen_attr);
+    if (status != UCS_OK) {
+        printf("<Failed to query UCP listener>\n");
+        goto out_destroy_listener;
+    }
+
+    status = ucs_sockaddr_get_port((struct sockaddr*)&listen_attr.sockaddr,
+                                   listen_port_p);
+    if (status != UCS_OK) {
+        printf("<Failed to get port>\n");
+        goto out_destroy_listener;
+    }
+
+    *listener_p = listener;
+out:
+    return status;
+
+out_destroy_listener:
+    ucp_listener_destroy(listener);
+    goto out;
+}
+
+ucs_status_t
+print_ucp_ep_info(ucp_worker_h worker, const ucp_ep_params_t *base_ep_params,
+                  const char *ip_addr)
+{
+    ucp_listener_h listener = NULL;
+    ucp_ep_h server_ep      = NULL;
+    ucs_status_t status, tmp_status;
+    ucs_status_ptr_t status_ptr;
+    ucp_ep_params_t ep_params;
+    ucp_address_t *address;
+    size_t address_length;
+    struct sockaddr_in connect_saddr;
+    uint16_t listen_port;
+    ucp_ep_h ep;
+    ucp_request_param_t request_param;
+
+    ep_params = *base_ep_params;
+
+    if (strlen(ip_addr) != 0) {
```

Comment:
maybe pass NULL instead of empty string?

Reply:
done